### PR TITLE
feat(document): add `$inc()` helper that increments numeric paths

### LINKED ISSUE #11915: can I do $inc while using .save?
### Prerequisites

- [X] I have written a descriptive issue title

### Mongoose version

5.13.6

### Node.js version

14.19.1

### MongoDB version

4.0.5

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

20.04

### Issue

I have a operation where an object is pushed into an array and a counter is increased.
```javascript
let foo = await Foo.findOne({ <some-filter> });
foo.bars.push({ bar: 2 });
foo.bar_count++;
await foo.save();
```
when I check the query being executed in debug mode, I see it is doing $push on array but $set on the count. In my use case I want to use $inc due to some concurrency issues. I searched and found `.$inc()` but that seems to be deprecated.
Is there any way by which I can tell `save` how to make the query? or maybe change the query which will be executed?

### PR DESCRIPTION
**Pull request by vkarpov15**
Fix #11915

**Summary**

See #11915. Right now, there's no way to use [`$inc`](https://www.mongodb.com/docs/manual/reference/operator/update/inc/) with `save()`. `doc.num += 2` becomes a `$set`.

With this change, you can explicitly increment a given path, and when you `save()`, Mongoose will send a `$inc`.

**Examples**

```javascript
const schema = new Schema({
  counter: Number
});
Test = db.model('Test', schema);

await Test.create({ counter: 0 });
const doc = await Test.findOne();
assert.strictEqual(doc.counter, 0);
const doc2 = await Test.findOne();

doc2.counter = 1;
await doc2.save();

doc.$inc('counter');
await doc.save();

const res = await Test.findById(doc);
assert.equal(res.counter, 2);
```

### CODE REVIEW

File: test/document.test.js [hunk 1]
```diff
@@ -11551,6 +11551,202 @@ describe('document', function() {
...
+      it('sends a $inc command for a given path', async function() {
+        await Test.create({ counter: 0 });
+        const doc = await Test.findOne();
+        assert.strictEqual(doc.counter, 0);
+        const doc2 = await Test.findOne();
+
+        doc2.counter = 1;
+        await doc2.save();
+
+        doc.$inc('counter');
```
Comment (AbdelrahmanHafez) [inline test/document.test.js:11575]: Would be nice to have the test explicitly increment specific value, to have some idea how it works when looking at the tests as an example:
```suggestion
        doc.$inc('counter', 1);
```

**Review by AbdelrahmanHafez — APPROVED** [review]
One minor nitpick, otherwise LGTM. 👍

Kind of surprised by the number of scenarios we have to take into account for this to work properly.

**Comment by AbdelrahmanHafez** [comment]
On a second thought, what do you think about the following syntax? How easy would it be to support it?
```js
document.counter.$inc(1);
document.friends[3].age.$inc(1);
```

**Comment by vkarpov15** [comment]
@AbdelrahmanHafez `document.counter.$inc(1);` is unfortunately not viable. [You can't define properties on primitives, it throws an error in strict mode](https://masteringjs.io/tutorials/fundamentals/strict#no-setting-properties-on-primitives).

**Comment by vkarpov15** [comment]
Re: number of cases, it is mostly just because I'm trying to be more defensive about shipping new features. In the past, we've often had edge cases in new features like "oh this doesn't work with subdocuments". So this time I'm trying to cover every potential bug I can think of with tests.

**Review by Uzlopak — CHANGES_REQUESTED** [review]
LGTM

**Review by Uzlopak — APPROVED** [review]
LGTM
